File: src/ambilight/AmbilightPreview.java
package ambilight;

import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;
import java.util.List;

import static java.lang.Math.*;

public class AmbilightPreview extends Ambilight {

	private final LedConfig config;
	private final byte[][] colors;
	private final List<Light> lights;

	private static final int lightCount = 5;
	private static final double dim = 1.5;

	public AmbilightPreview(LedConfig config) {
		this.config = config;

		colors = new byte[config.getLedCount()][];
		for (int i = 0; i < config.getLedCount(); i++) {
			colors[i] = new byte[3];
		}

		lights = new ArrayList<>();
		for (int i = 0; i < lightCount; i++) {
			lights.add(new Light(config.getLedsWidth(), config.getLedsHeight()));
		}
	}

	@NotNull
	@Override
	public byte[][] getScreenSegmentsColors() {
		for (int i = 0; i < config.getLedCount(); i++) {
			int[] led = config.getLed(i);
			int ledX = led[0];
			int ledY = led[1];

			int r = 0, g = 0, b = 0;
			for (Light light : lights) {
				double distX = light.x - ledX;
				double distY = light.y - ledY;
				double distance = sqrt(distX*distX + distY*distY);
				double distanceMulti = max(0, (light.size-distance) / light.size);
				r += light.r * distanceMulti;
				g += light.g * distanceMulti;
				b += light.b * distanceMulti;
			}

			colors[i][0] = (byte) min(255, r / dim);
			colors[i][1] = (byte) min(255, g / dim);
			colors[i][2] = (byte) min(255, b / dim);
		}

		return colors;
	}

	private class Light {

		double x, y;
		double velX, velY;
		double size;
		int r, g, b;

		Light(int ledWidth, int ledHeight) {
			r = (int) (256 * random());
			g = (int) (256 * random());
			b = (int) (256 * random());
			x = random() * ledWidth;
			y = random() * ledHeight;
			velX = random();
			velY = random();
			size = (0.5 + random()) * min(ledWidth, ledHeight);

			System.out.println("Color" +
							   " | x:" + (int) x +
							   " | y:" + (int) y +
							   " | size:" + (int) size);
		}
	}
}

File: core/src/main/java/me/khol/ambilight/AmbilightPreview.kt
package me.khol.ambilight

import java.lang.Math.*
import kotlin.math.min

class AmbilightPreview(private val config: LedConfig) : Ambilight() {

	companion object {
		private const val dim = 1.5f
	}

	private val colors = Array(config.ledCount) { LedColor() }
	private val lights = List(5) { Light(config.ledsWidth, config.ledsHeight) }

	override fun getScreenSegmentsColors(): Array<LedColor> {
		for (i in 0 until config.ledCount) {
			val (ledX, ledY) = config[i]

			var r = 0f
			var g = 0f
			var b = 0f
			for (light in lights) {
				val distX = light.x - ledX
				val distY = light.y - ledY
				val distance = sqrt(distX * distX + distY * distY)
				val distanceMulti = max(0.0, (light.size - distance) / light.size)
				r += (light.r * distanceMulti).toInt()
				g += (light.g * distanceMulti).toInt()
				b += (light.b * distanceMulti).toInt()
			}

			colors[i].r = (r / dim).coerceAtMost(255.0f)
			colors[i].g = (g / dim).coerceAtMost(255.0f)
			colors[i].b = (b / dim).coerceAtMost(255.0f)
		}

		return colors
	}

	private inner class Light(ledWidth: Int, ledHeight: Int) {

		internal var x = 0.0
		internal var y = 0.0
		internal var velX = 0.0
		internal var velY = 0.0
		internal var size = 0.0
		internal var r = 0
		internal var g = 0
		internal var b = 0

		init {
			r = (256 * random()).toInt()
			g = (256 * random()).toInt()
			b = (256 * random()).toInt()
			x = random() * ledWidth
			y = random() * ledHeight
			velX = random()
			velY = random()
			size = (0.5 + random()) * min(ledWidth, ledHeight)

			println("Color" +
				" | x:" + x.toInt() +
				" | y:" + y.toInt() +
				" | size:" + size.toInt())
		}
	}
}
